File: 2024/src/02/day02b.cc
#include <bits/stdc++.h>
using namespace std;

bool isSafe(vector<int> numbers)
{
    vector<int> dx(numbers.size() - 1, 0);
    bool incValid = true;
    bool decValid = true;

    for (int i = 0; i < numbers.size() - 1; i++)
        dx[i] = numbers[i + 1] - numbers[i];

    for (const auto &dxi : dx)
    {
        if (dxi < 1 || dxi > 3)
            incValid = false;
        if (dxi > -1 || dxi < -3)
            decValid = false;
    }
    return incValid || decValid;
}

int main()
{
    ifstream file{"../input/day02.in"};
    string line;
    int valids{0};
    while (getline(file, line))
    {
        vector<int> numbers;
        stringstream ss(line);
        int number{0};
        while (ss >> number)
        {
            numbers.push_back(number);
        }
        if (isSafe(numbers))
            valids++;
        else
        {
            for (int i = 0; i < numbers.size(); i++)
            {
                vector<int> numbersAlt = numbers;
                numbersAlt.erase(numbersAlt.begin() + i);
                if (isSafe(numbersAlt))
                {
                    valids++;
                    break;
                }
            }
        }
    }
    cout << valids << endl;
}

File: 2024/src/02/day02b.js
const fs = require("fs");

let valids = 0;
fs.readFile("input/day02.in", "utf8", (err, data) => {
  if (err) throw err;
  const lines = data.split("\n");
  lines.forEach((line) => {
    numbers = line.trim().split(/[\s,]+/);
    if (isSafe(numbers)) valids++;
    else {
      // Try removing any number and check if its safe
      for (let i = 0; i < numbers.length; i++) {
        let numbersAlt = numbers.slice();
        numbersAlt.splice(i, 1);
        if (isSafe(numbersAlt)) {
          valids++;
          break;
        }
      }
    }
  });
  console.log(valids);
});

function isSafe(numbers) {
  let dx = [];
  for (let i = 0; i < numbers.length - 1; i++) {
    dx.push(numbers[i + 1] - numbers[i]);
  }
  let incValid = true;
  let decValid = true;
  for (let i = 0; i < dx.length; i++) {
    if (dx[i] < 1 || dx[i] > 3) incValid = false;
    if (dx[i] > -1 || dx[i] < -3) decValid = false;
  }
  return incValid || decValid;
}
